hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/JetSqlBackend.java

```diff
@@ -0,0 +1,308 @@
+
+        return new CreateJobPlan(
+                sqlCreateJob.name(),
+                sqlCreateJob.jobConfig(),
+                sqlCreateJob.ifNotExists(),
+                dmlPlan,
+                planExecutor
+        );
+    }
+
+    private SqlPlan toAlterJobPlan(SqlAlterJob sqlAlterJob) {
+        return new AlterJobPlan(sqlAlterJob.name(), sqlAlterJob.getOperation(), planExecutor);
+    }
+
+    private SqlPlan toDropJobPlan(SqlDropJob sqlDropJob) {
+        return new DropJobPlan(sqlDropJob.name(), sqlDropJob.ifExists(), sqlDropJob.withSnapshotName(), planExecutor);
+    }
+
+    private SqlPlan toCreateSnapshotPlan(SqlCreateSnapshot sqlNode) {
+        return new CreateSnapshotPlan(sqlNode.getSnapshotName(), sqlNode.getJobName(), planExecutor);
+    }
+
+    private SqlPlan toDropSnapshotPlan(SqlDropSnapshot sqlNode) {
+        return new DropSnapshotPlan(sqlNode.getSnapshotName(), sqlNode.isIfExists(), planExecutor);
+    }
+
+    private ExecutionPlan toPlan(RelNode rel, List<String> fieldNames, OptimizerContext context) {
+        logger.fine("Before logical opt:\n" + RelOptUtil.toString(rel));
+        LogicalRel logicalRel = optimizeLogical(context, rel);
+        logger.fine("After logical opt:\n" + RelOptUtil.toString(logicalRel));
+        PhysicalRel physicalRel = optimizePhysical(context, logicalRel);
+        logger.fine("After physical opt:\n" + RelOptUtil.toString(physicalRel));
+
+        boolean isStreaming = containsStreamSource(rel);
+        boolean isInsert = physicalRel instanceof TableModify;
+
+        if (isInsert) {
+            DAG dag = createDag(physicalRel);
+            return new ExecutionPlan(dag, isStreaming, true, null, null, planExecutor);
+        } else {
+            QueryId queryId = QueryId.create(nodeEngine.getLocalMember().getUuid());
+            DAG dag = createDag(new JetRootRel(physicalRel, nodeEngine.getThisAddress(), queryId));
+            SqlRowMetadata rowMetadata = createRowMetadata(fieldNames, physicalRel.schema().getTypes());
+            return new ExecutionPlan(dag, isStreaming, false, queryId, rowMetadata, planExecutor);
+        }
+    }
+
+    /**
+     * Perform logical optimization.
+     *
+     * @param rel Original logical tree.
+     * @return Optimized logical tree.
+     */
+    private LogicalRel optimizeLogical(OptimizerContext context, RelNode rel) {
+        return (LogicalRel) context.optimize(
+                rel,
+                LogicalRules.getRuleSet(),
+                OptUtils.toLogicalConvention(rel.getTraitSet())
+        );
+    }
+
+    /**
+     * Perform physical optimization.
+     * This is where proper access methods and algorithms for joins and aggregations are chosen.
+     *
+     * @param rel Optimized logical tree.
+     * @return Optimized physical tree.
+     */
+    private PhysicalRel optimizePhysical(OptimizerContext context, RelNode rel) {
```

Comment:
When looking at the physical part I was surprised that you do not use distribution during planning. For example, it seems that the query like `SELECT * FROM VALUES (1) (2)` will be executed in distributed mode. Is this the case?

Reply:
Yes. Jet doesn't have means to dispatch some processors to some members. It always goes to all members, but some members will do nothing.